Add `ase` class as subdomain of `pr.create.structure`.

zendegani opened the pull request:
- `ase.build.cut` is available as `pr.create.structure.ase.cut` or `pr.create.structure.cut`
- `ase.build.stack` is available as `pr.create.structure.ase.stack` or `pr.create.structure.stack`
- `ase.spacegroup.crystal` is available as `pr.create.structure.ase.crystal`

Note: as @niklassiemer already mentioned in [#1307](https://github.com/pyiron/pyiron/issues/1307) there is [crystal](https://github.com/pyiron/pyiron_atomistics/blob/859dec1f2d027b51676726c8670e599c5a81b487/pyiron_atomistics/atomistics/structure/factory.py#L155) method in [StructureFactory](https://github.com/pyiron/pyiron_atomistics/blob/859dec1f2d027b51676726c8670e599c5a81b487/pyiron_atomistics/atomistics/structure/factory.py#L50). Therefore, `pr.create.structure.crystal` does not refer to `pr.create.structure.ase.crystal`.

zendegani committed:
import ase: crystal, cut, stack

zendegani committed:
Add ase as subdomain of pr.create.structure; Add ase.build.cut as pr.create.structure.ase.cut and pr.create.structure.cut

zendegani committed:
Add pr.create.structure.ase.stack and its wrapper pr.create.structure.stack

zendegani committed:
Fix typo

zendegani committed:
Add pr.create.structure.ase.crystal but didn't its wrapper because there is already a crystal method in StructureFactory

zendegani committed:
Merge branch 'master' into ase_crystal-cut-stack

jan-janssen committed:
Refactoring ofr ASE Factory

zendegani commented:
The main reason that I brought up this pull request during the meeting is that I had to implement `pr.create.structure.cut` as a `classmethod` and consequently it does not pass the docstrings.
Otherwise both `pr.create.structure.ase.cut` and `pr.create.structure.cut` are working properly as `ase.build.cut`.

There were couple of interesting comments during the meeting about how to refactor methods to improve the issue, yet when I tried them I've encountered another problems.
Therefore, it would be helpful if you implement your idea and try afterwards to run the following code to see if `pr.create.structure.cut` can also pass docstrings:

```python
from pyiron import Project
pr = Project('test')

el1, el2 = 'Fe', 'Nb'
a_c14, c_c14 = 4.77, 7.78
ase_c14 = pr.create.structure.ase.crystal(
    2*[el1]+[el2], 
    basis=[(0,0,0),(0.1697,0.3394,.25),(1/3.,2/3.,.5629)], 
    spacegroup=194, cellpar=[a_c14, a_c14, c_c14, 90., 90., 120.])
a_mu, c_mu =  a_c14, 26.78
ase_mu = pr.create.structure.ase.crystal(
    2*[el1]+3*[el2], 
    basis=[(0,0,0.5),(0.5,0.5,0.09),(0.,0.,0.05),(0.,0.,0.154),(0.,0.,1/3.)],
    spacegroup=166, cellpar=[a_mu, a_mu, c_mu, 90., 90., 120.])
# cut out the required layers
ase_mu_cross = pr.create.structure.cut(
    ase_mu, (1, 0, 0), (0, 1, 0), origo=(0, 0, .29), nlayers=3)
ase_lower = pr.create.structure.cut(
    ase_c14, (1, 0, 0), (0, 1, 0), nlayers=7)
ase_upper = pr.create.structure.cut(
    ase_c14.repeat((1,1,2)), (1, 0, 0), (0, 1, 0),
    origo=(0, 0, .1), nlayers=13)
#stacking the layers
ase_stack_low_mid = pr.create.structure.stack(ase_lower,ase_mu_cross, distance=2.5);
ase_stack_whole = pr.create.structure.stack(ase_stack_low_mid,ase_upper, distance=2.5);
cbd = ase_stack_whole.repeat((3,3,2))
cbd.plot3d()
```
```python
print(pr.create.structure.cut.__doc__)
```
```python
# print(pr.create.structure.ase.cut.__doc__)
```

jan-janssen commented:
@zendegani I solved this in https://github.com/pyiron/pyiron_atomistics/pull/83

liamhuber committed:
Avoid all fanciness

liamhuber commented:
> @zendegani I solved this in #83

@jan-janssen, when I ran Ali's example I still got the following error with your patch

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-74b5498d2a91> in <module>
     14     spacegroup=166, cellpar=[a_mu, a_mu, c_mu, 90., 90., 120.])
     15 # cut out the required layers
---> 16 ase_mu_cross = pr.create.structure.cut(
     17     ase_mu, (1, 0, 0), (0, 1, 0), origo=(0, 0, .29), nlayers=3)
     18 ase_lower = pr.create.structure.cut(

~/Liam/pyiron/pyiron_repos/pyiron_atomistics/pyiron_atomistics/atomistics/structure/factory.py in cut(cls, *args, **kwargs)
     99             pyiron_atomistics.atomstic.structure.ase.cut
    100         """
--> 101         return cls.ase.cut(*args, **kwargs)
    102 
    103     # def cut(*args, **kwargs):

AttributeError: type object 'StructureFactory' has no attribute 'ase'
```

I took the liberty of just directly pushing to your patch branch. @zendegani take a look at #83 again; it's running for me and the docstrings come out as I would hope. It was just a matter of purging all references to `@classmethod`.

sudarsan-surendralal commented on pyiron_atomistics/atomistics/structure/factory.py:
```diff
@@ -51,6 +53,64 @@
 
 
 class StructureFactory(PyironFactory):
+    class ase:
```
```suggestion
    class AseActions: #or something like that
```

sudarsan-surendralal commented on pyiron_atomistics/atomistics/structure/factory.py:
```diff
@@ -51,6 +53,64 @@
+
+            """
+            s.publication_add(publication_ase())
+            return ase_to_pyiron(ase_cut(*args, **kwargs))
```
In principle, this should work even if we pass a pyiron `Atoms` object to the function since it is now derived from `ase`. So there is no need for `ase_to_pyiron` (which can be slow, see: https://github.com/pyiron/pyiron_atomistics/issues/48). 

```suggestion
            return ase_cut(*args, **kwargs)
```

niklassiemer replied:
Yes, the ase `cut` and `stack` methods work with our structure class.

sudarsan-surendralal reviewed (commented):
These are very useful @zendegani! Here are some suggestions

zendegani committed:
Replace __doc__ addition with wraps decorator to pass also the signature

zendegani committed:
Remove ase_to_pyiron() from cut() and stack()

niklassiemer commented:
Actually, PR #83 wants to merge into this PR - so, just merge it in here and proceed with this one.

zendegani commented:
@niklassiemer you are right.

zendegani committed:
Merge pull request #83 from pyiron/ase-crystal-suggestion

Refactoring of ASEFactory

zendegani commented:
@jan-janssen and @liamhuber 
Very nice! It works perfectly.
I've done two minor changes though as following:
1. Replaced `__doc__` addition with `@wraps` decorator to pass also the `signature`, which I believe is a nice-to-have when getting `help()`.
2. Follwoing @sudarsan-surendralal [comment](https://github.com/pyiron/pyiron_atomistics/pull/76#discussion_r579790057), removed the `ase_to_pyiron `  from `cut` and `stack`, but not from `crystal`.

jan-janssen reviewed (approved):
Looks good to me
